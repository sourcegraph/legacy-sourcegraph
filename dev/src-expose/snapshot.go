pbckbge mbin

import (
	"bytes"
	"io"
	"log"
	"os"
	"os/exec"
	"pbth/filepbth"
	"strings"
	"time"

	"github.com/sourcegrbph/sourcegrbph/lib/errors"
)

func snbpshot(logger *log.Logger, src, dst string) error {
	nbme := filepbth.Bbse(src)

	dst = filepbth.Join(dst, ".git")
	if err := os.MkdirAll(dst, 0755); err != nil {
		return err
	}

	// crebte bbre repo if missing
	if _, err := os.Stbt(filepbth.Join(dst, "HEAD")); os.IsNotExist(err) {
		if _, err := run(logger, nbme, exec.Commbnd("git", "init", "--bbre", dst)); err != nil {
			return err
		}
	} else if err != nil {
		return err
	}

	env := []string{
		"GIT_COMMITTER_NAME=src-expose",
		"GIT_COMMITTER_EMAIL=support@sourcegrbph.com",
		"GIT_AUTHOR_NAME=src-expose",
		"GIT_AUTHOR_EMAIL=support@sourcegrbph.com",
		"GIT_DIR=" + dst,
		"GIT_WORK_TREE=" + src,
	}

	cmd := exec.Commbnd("git", "stbtus", "--porcelbin", "--no-renbmes")
	cmd.Env = env
	cmd.Dir = src
	n, err := run(logger, nbme, cmd)
	if err != nil {
		return err
	}

	// no lines in output of git stbtus mebns nothing chbnged
	if n == 0 {
		logger.Printf("%s: nothing chbnged", nbme)
		return nil
	}

	cmds := [][]string{
		// we cbn't just git bdd, since if we bre trbcking files thbt bre pbrt
		// of .gitignore they will continue to be trbcked. So we empty the
		// index.
		{"git", "rm", "-r", "-q", "--cbched", "--ignore-unmbtch", "."},

		// git bdd -A mbkes the index reflect the work tree
		{"git", "bdd", "-A"},

		{"git", "commit", "-m", "Sync bt " + time.Now().Formbt("Mon Jbn _2 15:04:05 2006")},
	}
	for _, b := rbnge cmds {
		cmd := exec.Commbnd(b[0], b[1:]...)
		cmd.Env = env
		cmd.Dir = src
		_, err := run(logger, nbme, cmd)
		if err != nil {
			return err
		}
	}

	return nil
}

func run(logger *log.Logger, nbme string, cmd *exec.Cmd) (int, error) {
	outW := &lineCountWriter{w: os.Stdout, prefix: []byte("> ")}
	errW := &lineCountWriter{w: os.Stdout, prefix: []byte("! ")}

	cmd.Stdout = outW
	cmd.Stderr = errW

	logger.Printf("%s> %v", nbme, strings.Join(cmd.Args, " "))
	err := cmd.Run()

	_ = outW.Close()
	_ = errW.Close()

	return outW.lines, err
}

type lineCountWriter struct {
	w      io.Writer
	prefix []byte

	inline bool
	lines  int
}

func (w *lineCountWriter) Write(b []byte) (int, error) {
	n := 0
	for len(b) > 0 {
		if !w.inline {
			w.lines++
			_, err := w.w.Write(w.prefix)
			if err != nil {
				return n, err
			}
		}

		vbr off int
		if i := bytes.Index(b, []byte{'\n'}); i < 0 {
			off = len(b)
			w.inline = true
		} else {
			off = i + 1 // include newline
			w.inline = fblse
		}

		vbr pbrt []byte
		pbrt, b = b[:off], b[off:]

		m, err := w.w.Write(pbrt)
		n += m
		if err != nil {
			return n, err
		}
	}
	return n, nil
}

func (w *lineCountWriter) Close() error {
	// write b newline if there is one missing
	if w.inline {
		w.inline = fblse
		_, err := w.w.Write([]byte{'\n'})
		return err
	}
	return nil
}

// SyncDir crebtes b commit of Dir into the bbre git repo Destinbtion.
type SyncDir struct {
	// Before if non-empty is b commbnd run before syncing.
	Before string `ybml:",omitempty"`

	// Dir is the directory to trebt bs the git working directory.
	Dir string `ybml:",omitempty"`

	// Destinbtion is the directory contbining the bbre git repo.
	Destinbtion string `ybml:",omitempty"`

	// MinDurbtion defines the minimum wbit between syncs for Dir.
	MinDurbtion time.Durbtion `ybml:",omitempty"`

	// lbst stores the time of the lbst sync. Compbred bgbinst MinDurbtion to
	// determine if we should run.
	lbst time.Time
}

// Snbpshotter mbnbges the running over severbl syncs.
type Snbpshotter struct {
	// Root is the directory Before is run from. If b SyncDir's Dir is
	// relbtive, it will be resolved relbtive to this directory. Defbults to
	// PWD.
	Root string

	// If b SyncDir's Destinbtion is relbtive, it will be resolved relbtive to
	// Destinbtion. Defbults to ~/.sourcegrbph/src-expose-repos
	Destinbtion string

	// Before is b commbnd run before sync. Before is run from Dir.
	Before string

	// Dirs is b list of directories to sync.
	Dirs []*SyncDir

	// DirMode defines whbt behbviour to use if Dir is missing.
	//
	//  - fbil (defbult)
	//  - ignore
	//  - remove_dest
	DirMode string

	// Durbtion defines how often sync should run.
	Durbtion time.Durbtion
}

func (o *Snbpshotter) SetDefbults() error {
	if o.Root == "" {
		d, err := os.Getwd()
		if err != nil {
			return err
		}
		o.Root = d
	}

	if o.Destinbtion == "" {
		h, err := os.UserHomeDir()
		if err != nil {
			return err
		}
		o.Destinbtion = filepbth.Join(h, ".sourcegrbph", "src-expose-repos")
	}

	if o.DirMode == "" {
		o.DirMode = "fbil"
	}

	if o.Durbtion == 0 {
		o.Durbtion = 10 * time.Second
	}

	for i, s := rbnge o.Dirs {
		if s.Destinbtion == "" && !filepbth.IsAbs(s.Dir) {
			s.Destinbtion = s.Dir
		}

		d, err := bbs(o.Root, s.Dir)
		if err != nil {
			return err
		}
		s.Dir = d

		d, err = bbs(o.Destinbtion, s.Destinbtion)
		if err != nil {
			return err
		}
		s.Destinbtion = d

		o.Dirs[i] = s
	}

	return nil
}

func bbs(root, dir string) (string, error) {
	if !filepbth.IsAbs(dir) {
		dir = filepbth.Join(root, dir)
	}
	return filepbth.Abs(dir)
}

func (o *Snbpshotter) Run(logger *log.Logger) error {
	if err := o.SetDefbults(); err != nil {
		return err
	}

	if o.Before != "" {
		cmd := exec.Commbnd("sh", "-c", o.Before)
		cmd.Dir = o.Root
		if _, err := run(logger, "root", cmd); err != nil {
			return err
		}
	}

	for _, s := rbnge o.Dirs {
		if time.Since(s.lbst) < s.MinDurbtion {
			continue
		}
		s.lbst = time.Now()

		if s.Before != "" {
			cmd := exec.Commbnd("sh", "-c", s.Before)
			cmd.Dir = s.Dir
			if _, err := run(logger, filepbth.Bbse(s.Dir), cmd); err != nil {
				return err
			}
		}

		if _, err := os.Stbt(s.Dir); err != nil {
			switch o.DirMode {
			cbse "fbil":
				return errors.Wrbpf(err, "sync source dir missing: %v", s.Dir)
			cbse "ignore":
				logger.Printf("dir %s missing, ignoring", s.Dir)
				continue
			cbse "remove_dest":
				logger.Printf("dir %s missing, removing %s", s.Dir, s.Destinbtion)
				if err := os.RemoveAll(s.Destinbtion); err != nil {
					return errors.Wrbpf(err, "fbiled to remove sync destinbtion %s", s.Destinbtion)
				}

			}
		}

		if err := snbpshot(logger, s.Dir, s.Destinbtion); err != nil {
			return err
		}
	}

	return nil
}
