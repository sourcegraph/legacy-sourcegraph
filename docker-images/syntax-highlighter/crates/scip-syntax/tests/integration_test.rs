use std::{
    collections::{HashMap, HashSet},
    path::{Path, PathBuf},
    process::Command,
};

use assert_cmd::{cargo::cargo_bin, prelude::*};
use scip_syntax::{
    evaluate::Evaluator,
    index::{index_command, AnalysisMode, IndexMode, IndexOptions},
    io::read_index_from_file,
};

lazy_static::lazy_static! {
    static ref BINARY_LOCATION: PathBuf = {
        match std::env::var("SCIP_SYNTAX_PATH") {
            Ok(va) => std::env::current_dir().unwrap().join(va),
            _ => cargo_bin("scip-syntax"),
        }
    };

    static ref BASE: PathBuf = {
        match std::env::var("CARGO_MANIFEST_DIR") {
            Ok(va) => std::env::current_dir().unwrap().join(va),
            _ => std::env::current_dir().unwrap()            }
    };

    static ref JAVA_SCIP_INDEX: PathBuf = {
        match std::env::var("JAVA_SCIP_INDEX") {
            Ok(va) => std::env::current_dir().unwrap().join(va),
            _ => BASE.join("testdata/java/index.scip")
        }
    };


}

use syntax_analysis::snapshot::{dump_document_with_config, EmitSymbol, SnapshotOptions};

fn snapshot_syntax_document(doc: &scip::types::Document, source: &str) -> String {
    dump_document_with_config(
        doc,
        source,
        SnapshotOptions {
            emit_symbol: EmitSymbol::All,
            ..Default::default()
        },
    )
    .expect("dump document")
}

#[test]
fn java_e2e_evaluation() {
    let dir = BASE.join("testdata/java");

    let out_dir = tempdir();

    let candidate = out_dir.join("index-tree-sitter.scip");

    index_command(
        "java".to_string(),
        IndexMode::Workspace {
            location: dir.clone(),
        },
        candidate.clone(),
        dir.clone(),
        None,
        IndexOptions {
            analysis_mode: AnalysisMode::Full,
            fail_fast: true,
        },
    )
    .unwrap();

    let mut str = vec![];

    Evaluator::default()
        .evaluate_files(candidate, JAVA_SCIP_INDEX.to_path_buf())
        .unwrap()
        .write_summary(
            &mut str,
            scip_syntax::evaluate::EvaluationOutputOptions {
                print_false_negatives: true,
                print_true_positives: true,
                print_false_positives: true,
                print_mapping: true,
                disable_colors: true,
            },
        )
        .unwrap();

    insta::assert_snapshot!("java_evaluation", String::from_utf8(str).unwrap());
}

#[test]
fn java_e2e_indexing() {
    let out_dir = tempdir();
    let setup = HashMap::from([(
        PathBuf::from("globals.java"),
        include_str!("../testdata/globals.java").to_string(),
    )]);

    let mut cmd = command("index");
    let output_location = out_dir.join("index.scip");
    let paths: Vec<String> = setup
        .clone()
        .into_keys()
        .map(|pb| pb.to_str().unwrap().to_string())
        .collect();

    prepare(&out_dir, &setup);

    cmd.args(vec![
        "--language",
        "java",
        "--out",
        output_location.to_str().unwrap(),
    ])
    .current_dir(out_dir)
    .args(paths)
    .assert()
    .success();

    let index = read_index_from_file(&output_location).unwrap();
    let mut indexed_paths: HashSet<String> = HashSet::new();

    for doc in &index.documents {
        let path = &doc.relative_path;
        indexed_paths.insert(path.to_string());
        let dumped = snapshot_syntax_document(
            doc,
            setup.get(&PathBuf::from(&path)).expect(
                format!(
                    "Unexpected relative path {} found in the index. Valid paths are: {:?}",
                    path,
                    setup.keys()
                )
                .as_str(),
            ),
        );

        insta::assert_snapshot!(format!("files_indexing-{}", path.clone()), &dumped);
    }

    assert_eq!(
        indexed_paths,
        setup
            .into_keys()
            .map(|p| p.to_str().unwrap().to_string())
            .collect()
    );
}

#[test]
fn java_workspace_indexing() {
    let out_dir = tempdir();
    let setup = HashMap::from([
        (
            PathBuf::from("src/main/java/globals.java"),
            include_str!("../testdata/globals.java").to_string(),
        ),
        (
            PathBuf::from("package-info.java"),
            include_str!("../testdata/package-info.java").to_string(),
        ),
    ]);

    let mut cmd = command("index");
    let output_location = out_dir.join("index.scip");

    prepare(&out_dir, &setup);

    cmd.args(vec![
        "--workspace",
        out_dir.to_str().unwrap(),
        "--language",
        "java",
        "--out",
        output_location.to_str().unwrap(),
    ])
    .assert()
    .success();

    let index = read_index_from_file(&output_location).unwrap();

    let mut indexed_paths: HashSet<String> = HashSet::new();

    for doc in &index.documents {
        let path = &doc.relative_path;
        indexed_paths.insert(path.to_string());
        let dumped = snapshot_syntax_document(
            doc,
            setup.get(&PathBuf::from(&path)).expect(
                format!(
                    "Unexpected relative path {} found in the index. Valid paths are: {:?}",
                    path,
                    setup.keys()
                )
                .as_str(),
            ),
        );

        insta::assert_snapshot!(format!("workspace_indexing-{}", path.clone()), &dumped);
    }

    println!("{:?}", index.documents);

    assert_eq!(
        indexed_paths,
        setup
            .into_keys()
            .map(|p| p.to_str().unwrap().to_string())
            .collect()
    );
}

fn prepare(temp: &Path, files: &HashMap<PathBuf, String>) {
    for (path, contents) in files.iter() {
        let file_path = temp.join(path);
        write_file(&file_path, contents);
    }
}

fn command(sub: &str) -> Command {
    let mut cmd = Command::new(BINARY_LOCATION.to_str().unwrap());

    cmd.arg(sub);

    cmd
}

fn write_file(path: &PathBuf, contents: &String) {
    use std::io::Write;

    let Some(parent) = path.parent() else {
        panic!("failed to find parent dir for {:?}", path)
    };

    std::fs::create_dir_all(parent)
        .expect(format!("Failed to create all parent folders for {:?}", path).as_str());

    let output = std::fs::File::create(path)
        .expect(format!("Failed to open file {} for writing", path.to_str().unwrap()).as_str());
    let mut writer = std::io::BufWriter::new(output);
    writer.write_all(contents.as_bytes()).unwrap();
}

fn tempdir() -> PathBuf {
    tempfile::tempdir().unwrap().into_path()
}
