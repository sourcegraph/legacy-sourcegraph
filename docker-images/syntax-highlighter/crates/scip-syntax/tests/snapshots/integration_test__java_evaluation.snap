---
source: crates/scip-syntax/tests/integration_test.rs
expression: "String::from_utf8(str).unwrap()"
---
{"precision_percent":"96.3","recall_percent":"94.0","true_positives":"77.8","false_positives":"3.0","false_negatives":"5.0"}

False negatives (FN): 5
[FN] src/main/java/TodoList.java: L0 C8 -- semanticdb maven . . com/
[FN] src/main/java/TodoList.java: L0 C30 -- semanticdb maven . . com/sourcegraph/graph/twodwo/
[FN] src/main/java/TodoList.java: L2 C7 -- semanticdb maven . . java/
[FN] src/main/java/Main.java: L0 C8 -- semanticdb maven . . com/
[FN] src/main/java/Main.java: L0 C30 -- semanticdb maven . . com/sourcegraph/graph/twodwo/

False positives: 3
How many extra occurrences we reported compared to compiler?
[FP] src/main/java/TodoList.java: L0 C8 -- scip-syntax . . . `com.sourcegraph.graph.twodwo`/
[FP] src/main/java/TodoList.java: L2 C7 -- scip-syntax . . . `java.util`/
[FP] src/main/java/Main.java: L0 C8 -- scip-syntax . . . `com.sourcegraph.graph.twodwo`/

True positives: 87
[TP] src/main/java/TodoList.java: L0 C12 -- semanticdb maven . . com/sourcegraph/
[TP] src/main/java/TodoList.java: L0 C24 -- semanticdb maven . . com/sourcegraph/graph/
[TP] src/main/java/TodoList.java: L2 C12 -- semanticdb maven . . java/util/
[TP] src/main/java/TodoList.java: L2 C17 -- semanticdb maven jdk 17 java/util/ArrayList#
[TP] src/main/java/TodoList.java: L4 C13 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#
[TP] src/main/java/TodoList.java: L6 C21 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#
[TP] src/main/java/TodoList.java: L7 C10 -- semanticdb maven jdk 17 java/lang/String#
[TP] src/main/java/TodoList.java: L7 C17 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#title.
[TP] src/main/java/TodoList.java: L8 C18 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#done.
[TP] src/main/java/TodoList.java: L10 C9 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#`<init>`().
[TP] src/main/java/TodoList.java: L10 C14 -- semanticdb maven jdk 17 java/lang/String#
[TP] src/main/java/TodoList.java: L10 C21 -- local doc-0-0
[TP] src/main/java/TodoList.java: L11 C8 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#title.
[TP] src/main/java/TodoList.java: L11 C16 -- local doc-0-0
[TP] src/main/java/TodoList.java: L12 C8 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#done.
[TP] src/main/java/TodoList.java: L15 C9 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#`<init>`(+1).
[TP] src/main/java/TodoList.java: L15 C14 -- semanticdb maven jdk 17 java/lang/String#
[TP] src/main/java/TodoList.java: L15 C21 -- local doc-0-1
[TP] src/main/java/TodoList.java: L15 C36 -- local doc-0-2
[TP] src/main/java/TodoList.java: L16 C8 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#title.
[TP] src/main/java/TodoList.java: L16 C16 -- local doc-0-1
[TP] src/main/java/TodoList.java: L17 C8 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#done.
[TP] src/main/java/TodoList.java: L17 C15 -- local doc-0-2
[TP] src/main/java/TodoList.java: L20 C9 -- semanticdb maven jdk 17 java/lang/String#
[TP] src/main/java/TodoList.java: L20 C16 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#getTitle().
[TP] src/main/java/TodoList.java: L21 C15 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#title.
[TP] src/main/java/TodoList.java: L25 C9 -- semanticdb maven jdk 17 java/util/ArrayList#
[TP] src/main/java/TodoList.java: L25 C19 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#
[TP] src/main/java/TodoList.java: L25 C25 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#todos.
[TP] src/main/java/TodoList.java: L27 C8 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#`<init>`().
[TP] src/main/java/TodoList.java: L28 C2 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#todos.
[TP] src/main/java/TodoList.java: L28 C14 -- semanticdb maven jdk 17 java/util/ArrayList#`<init>`(+1).
[TP] src/main/java/TodoList.java: L31 C13 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#addTodo().
[TP] src/main/java/TodoList.java: L31 C21 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#
[TP] src/main/java/TodoList.java: L31 C26 -- local doc-0-3
[TP] src/main/java/TodoList.java: L32 C2 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#todos.
[TP] src/main/java/TodoList.java: L32 C8 -- semanticdb maven jdk 17 java/util/ArrayList#add(+1).
[TP] src/main/java/TodoList.java: L32 C12 -- local doc-0-3
[TP] src/main/java/TodoList.java: L35 C8 -- semanticdb maven jdk 17 java/util/ArrayList#
[TP] src/main/java/TodoList.java: L35 C18 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#
[TP] src/main/java/TodoList.java: L35 C24 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#getTodos().
[TP] src/main/java/TodoList.java: L36 C9 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#todos.
[TP] src/main/java/TodoList.java: L39 C13 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#summarise().
[TP] src/main/java/TodoList.java: L40 C4 -- semanticdb maven jdk 17 java/lang/System#
[TP] src/main/java/TodoList.java: L40 C11 -- semanticdb maven jdk 17 java/lang/System#out.
[TP] src/main/java/TodoList.java: L40 C15 -- semanticdb maven jdk 17 java/io/PrintStream#println(+8).
[TP] src/main/java/TodoList.java: L41 C7 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#
[TP] src/main/java/TodoList.java: L41 C12 -- local doc-0-4
[TP] src/main/java/TodoList.java: L41 C19 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#todos.
[TP] src/main/java/TodoList.java: L42 C7 -- local doc-0-4
[TP] src/main/java/TodoList.java: L42 C12 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#done.
[TP] src/main/java/TodoList.java: L43 C4 -- semanticdb maven jdk 17 java/lang/System#
[TP] src/main/java/TodoList.java: L43 C11 -- semanticdb maven jdk 17 java/lang/System#out.
[TP] src/main/java/TodoList.java: L43 C15 -- semanticdb maven jdk 17 java/io/PrintStream#println(+8).
[TP] src/main/java/TodoList.java: L43 C33 -- local doc-0-4
[TP] src/main/java/TodoList.java: L43 C38 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#getTitle().
[TP] src/main/java/TodoList.java: L45 C4 -- semanticdb maven jdk 17 java/lang/System#
[TP] src/main/java/TodoList.java: L45 C11 -- semanticdb maven jdk 17 java/lang/System#out.
[TP] src/main/java/TodoList.java: L45 C15 -- semanticdb maven jdk 17 java/io/PrintStream#println(+8).
[TP] src/main/java/TodoList.java: L45 C33 -- local doc-0-4
[TP] src/main/java/TodoList.java: L45 C38 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#getTitle().
[TP] src/main/java/Main.java: L0 C12 -- semanticdb maven . . com/sourcegraph/
[TP] src/main/java/Main.java: L0 C24 -- semanticdb maven . . com/sourcegraph/graph/
[TP] src/main/java/Main.java: L3 C13 -- semanticdb maven . . com/sourcegraph/graph/twodwo/Main#`<init>`().
[TP] src/main/java/Main.java: L5 C23 -- semanticdb maven . . com/sourcegraph/graph/twodwo/Main#main().
[TP] src/main/java/Main.java: L5 C28 -- semanticdb maven jdk 17 java/lang/String#
[TP] src/main/java/Main.java: L5 C37 -- local doc-1-0
[TP] src/main/java/Main.java: L7 C12 -- local doc-1-1
[TP] src/main/java/Main.java: L7 C27 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#`<init>`().
[TP] src/main/java/Main.java: L9 C12 -- semanticdb maven jdk 17 java/lang/String#
[TP] src/main/java/Main.java: L9 C19 -- local doc-1-2
[TP] src/main/java/Main.java: L9 C22 -- local doc-1-0
[TP] src/main/java/Main.java: L10 C12 -- local doc-1-1
[TP] src/main/java/Main.java: L10 C21 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#addTodo().
[TP] src/main/java/Main.java: L10 C33 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#
[TP] src/main/java/Main.java: L10 C42 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#`<init>`().
[TP] src/main/java/Main.java: L10 C47 -- local doc-1-2
[TP] src/main/java/Main.java: L13 C8 -- local doc-1-1
[TP] src/main/java/Main.java: L13 C17 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#addTodo().
[TP] src/main/java/Main.java: L13 C29 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#
[TP] src/main/java/Main.java: L13 C38 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#`<init>`(+1).
[TP] src/main/java/Main.java: L14 C8 -- local doc-1-1
[TP] src/main/java/Main.java: L14 C17 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#addTodo().
[TP] src/main/java/Main.java: L14 C29 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#
[TP] src/main/java/Main.java: L14 C38 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#Item#`<init>`().
[TP] src/main/java/Main.java: L17 C8 -- local doc-1-1
[TP] src/main/java/Main.java: L17 C17 -- semanticdb maven . . com/sourcegraph/graph/twodwo/TodoList#summarise().

Symbol mapping
scip-syntax . . . `com.sourcegraph.graph.twodwo`/TodoList#
   1.00 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList# [1/4 occurrences]

scip-syntax . . . `com.sourcegraph.graph.twodwo`/TodoList#summarise().
   1.00 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#summarise(). [1/2 occurrences]

scip-syntax . . . `com.sourcegraph.graph.twodwo`/TodoList#getTodos().
   1.00 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#getTodos(). [1/1 occurrences]

scip-syntax . . . `com.sourcegraph.graph.twodwo`/TodoList#addTodo().
   1.00 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#addTodo(). [1/4 occurrences]

scip-syntax . . . `com.sourcegraph.graph.twodwo`/TodoList#todos.
   1.00 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#todos. [1/5 occurrences]

scip-syntax . . . `com.sourcegraph.graph.twodwo`/TodoList#Item#
   1.00 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#Item# [1/5 occurrences]

scip-syntax . . . `com.sourcegraph.graph.twodwo`/TodoList#Item#getTitle().
   1.00 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#Item#getTitle(). [1/3 occurrences]

scip-syntax . . . `com.sourcegraph.graph.twodwo`/TodoList#Item#done.
   1.00 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#Item#done. [1/4 occurrences]

scip-syntax . . . `com.sourcegraph.graph.twodwo`/TodoList#Item#title.
   1.00 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#Item#title. [1/4 occurrences]

local doc-0-1
   1.00 local doc-0-0 [2/2 occurrences]

local doc-0-2
   1.00 local doc-0-1 [2/2 occurrences]

local doc-0-3
   1.00 local doc-0-2 [2/2 occurrences]

local doc-0-4
   1.00 local doc-0-3 [2/2 occurrences]

local doc-0-5
   1.00 local doc-0-4 [4/4 occurrences]

scip-syntax . . . sourcegraph.
   1.00 semanticdb . . . com/sourcegraph/ [2/2 occurrences]

scip-syntax . . . graph.
   1.00 semanticdb . . . com/sourcegraph/graph/ [2/2 occurrences]

scip-syntax . . . util.
   1.00 semanticdb . . . java/util/ [1/1 occurrences]

scip-syntax . . . ArrayList#
   0.50 semanticdb . . . java/util/ArrayList# [3/3 occurrences]
   0.50 semanticdb . . . java/util/ArrayList#`<init>`(+1). [1/1 occurrences]

scip-syntax . . . Item#
   0.34 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#Item#`<init>`(). [2/3 occurrences]
   0.25 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#Item#`<init>`(+1). [1/2 occurrences]
   0.41 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#Item# [4/5 occurrences]

scip-syntax . . . String#
   1.00 semanticdb . . . java/lang/String# [6/6 occurrences]

scip-syntax . . . Item.
   0.40 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#Item#`<init>`(). [1/3 occurrences]
   0.60 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#Item#`<init>`(+1). [1/2 occurrences]

scip-syntax . . . title.
   1.00 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#Item#title. [3/4 occurrences]

scip-syntax . . . done.
   1.00 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#Item#done. [3/4 occurrences]

scip-syntax . . . TodoList.
   1.00 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#`<init>`(). [1/2 occurrences]

scip-syntax . . . todos.
   1.00 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#todos. [4/5 occurrences]

scip-syntax . . . add().
   1.00 semanticdb . . . java/util/ArrayList#add(+1). [1/1 occurrences]

scip-syntax . . . System.
   1.00 semanticdb . . . java/lang/System# [3/3 occurrences]

scip-syntax . . . out.
   1.00 semanticdb . . . java/lang/System#out. [3/3 occurrences]

scip-syntax . . . println().
   1.00 semanticdb . . . java/io/PrintStream#println(+8). [3/3 occurrences]

scip-syntax . . . getTitle().
   1.00 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#Item#getTitle(). [2/3 occurrences]

scip-syntax . . . `com.sourcegraph.graph.twodwo`/Main#
   1.00 semanticdb . . . com/sourcegraph/graph/twodwo/Main#`<init>`(). [1/1 occurrences]

scip-syntax . . . `com.sourcegraph.graph.twodwo`/Main#main().
   1.00 semanticdb . . . com/sourcegraph/graph/twodwo/Main#main(). [1/1 occurrences]

local doc-1-1
   1.00 local doc-1-0 [2/2 occurrences]

local doc-1-2
   1.00 local doc-1-1 [5/5 occurrences]

local doc-1-3
   1.00 local doc-1-2 [2/2 occurrences]

scip-syntax . . . TodoList#
   0.40 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#`<init>`(). [1/2 occurrences]
   0.60 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList# [3/4 occurrences]

scip-syntax . . . addTodo().
   1.00 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#addTodo(). [3/4 occurrences]

scip-syntax . . . summarise().
   1.00 semanticdb . . . com/sourcegraph/graph/twodwo/TodoList#summarise(). [1/2 occurrences]


